File: 020. Selection Sort/SelectionSort.java
import java.util.*;

class SelectionSort {
    public static int[] selectionSort(int[] array) {
        if (array.length == 0)
            return array;

        int smallestIndex;
        for (int i = 0; i < array.length - 1; i++) {
            smallestIndex = i;
            for (int j = i + 1; j < array.length; j++) {
                if (array[j] < array[smallestIndex]) {
                    smallestIndex = j;
                }
            }
            if (array[smallestIndex] != array[i]) {
                swap(i, smallestIndex, array);
            }
        }
        return array;
    }

    public static void swap(int i, int j, int[] array) {
        int temp = array[i];
        array[i] = array[j];
        array[j] = temp;
    }
}

File: 020. Selection Sort/SelectionSort.kt
package com.algoexpert.program

fun selectionSort(array: MutableList<Int>): List<Int> {
    if (array.size == 0) return array
    var smallestIndex: Int
    for (i in 0 until array.size - 1) {
        smallestIndex = i
        for (j in i + 1 until array.size) {
            if (array[j] < array[smallestIndex]) {
                smallestIndex = j
            }
        }
        if (smallestIndex != i) {
            swap(i, smallestIndex, array)
        }
    }
    return array
}

fun swap(i: Int, j: Int, array: MutableList<Int>) {
    array[i] = array[j].also { array[j] = array[i] }
}
